### PR TITLE
Refacotring type generator

### PR DESCRIPTION
**Pull request by james-stephenson**
After working on #5 I spent some time working on the type generator so that more customizations could be added due to nuances in the Unifi API. The goal was to create a go text template file that could be customized for generating code, since in exposing the `Device` and `Settings*` APIs showed there were specific changes that needed to be made to support these types.

It's not perfect and there's a lot that could be improved but I felt this would be a worthwhile contribution for future maintenance.

I separated this PR into 3 individual commits to make the review much easier, I suggest reviewing them separately. The first commit does all the refactoring of the generator, which should make the second easier to review (splitting out sub types into top-level types within the generated code). The final commit makes the API current (5.14.23).

### CODE REVIEW

File: fields/main.go [hunk 1]
```diff
@@ -130,299 +221,185 @@ func main() {
...
-		return fmt.Sprintf("[]%s", elementType), elementComment, true, nil
+
+		fieldInfo.OmitEmpty = true
+		fieldInfo.IsArray = true
+
+		return fieldInfo, nil
 	case map[string]interface{}:
-		fieldNames := []string{}
-		fieldCodes := map[string]string{}
+		typeName := r.StructName + "_" + fieldName
```
Comment (paultyng) [inline fields/main.go:307]: I'd prefer to not use an underscore here, but if it results in conflicts its a reasonable compromise.

**Comment by paultyng** [comment]
@james-stephenson this mostly looks great to me, only the one comment above.

File: fields/main.go (hunk 1 above)
Comment (james-stephenson) [reply fields/main.go:307]: I agree with this. I added the underscore to ensure there were no type naming conflicts, but I don't know if it's necessary. If there's a future conflict that arises, we could instead create a special rule for just that type.

I'll remove this and do some more testing. Thanks!

**Comment by james-stephenson** [comment]
I found a few bugs with the Device generated API that I need to fixup, and will try to push that up soon so we can merge this PR.

**Comment by james-stephenson** [comment]
I've fixed up the tests to match, this should be ready for merge.

**Comment by paultyng** [comment]
LGTM, over time we will probably need to cleanup the initialism replacements to be more inline with [Go idiom](https://github.com/golang/go/wiki/CodeReviewComments#initialisms), which is technically breaking, but not something that will probably impact the provider much, so not too worried about it.

**Review by paultyng — APPROVED** [review]

**Pull request merged**